internal/security/secretstore/vmkencryption.go

```diff
@@ -0,0 +1,280 @@
+			v.logger.Error(fmt.Sprintf("failed to decode hex bytes of ciphertext"))
+			return err
+		}
+
+		keyShare, err := v.gcmDecryptKeyshare(cipherText, nonce, i) // Unwrap using a unique AES key
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to unwrap key %d: %s", i, err.Error()))
+			return err
+		}
+
+		newKeys[i] = hex.EncodeToString(keyShare)
+		newKeysBase64[i] = base64.StdEncoding.EncodeToString(keyShare)
+	}
+
+	initResp.Keys = newKeys
+	initResp.KeysBase64 = newKeysBase64
+	initResp.EncryptedKeys = nil
+	initResp.Nonces = nil
+	return nil
+}
+
+//
+// Internal methods
+//
+
+func (v *VMKEncryption) gcmEncryptKeyshare(keyshare []byte, counter int) ([]byte, []byte, error) {
+
+	info := fmt.Sprintf("vault%d", counter)
+
+	key, err := v.kdf.DeriveKey(v.ikm, aesKeyLength, info)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to derive encryption key for vault master key share %s", err.Error()))
+		return nil, nil, err
+	}
+	defer wipeKey(key)
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to initialize block cipher %s", err.Error()))
+		return nil, nil, err
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to initialize AES cipher %s", err.Error()))
+		return nil, nil, err
+	}
+
+	nonce := make([]byte, aesgcm.NonceSize())
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		v.logger.Error(fmt.Sprintf("failed to initialize random nonce %s", err.Error()))
+		return nil, nil, err
+	}
+
+	// Encrypt they key share, overwrite the input, and wipe the encrypted copy
+	// (input and output size are the same with AES)
+	ciphertext := aesgcm.Seal(nil, nonce, keyshare, nil)
+	wipeKey(keyshare)
+
+	return ciphertext, nonce, nil
+}
+
+func (v *VMKEncryption) gcmDecryptKeyshare(keyshare []byte, nonce []byte, counter int) ([]byte, error) {
+	// Each key shared is encrypted with a unique key
+	// by varying the info argument: vault0, vault1, etc...
+	info := fmt.Sprintf("vault%d", counter)
+
+	key, err := v.kdf.DeriveKey(v.ikm, aesKeyLength, info)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to derive encryption key for vault master key share %s", err.Error()))
+		return nil, err
+	}
+	defer wipeKey(key)
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to initialize block cipher %s", err.Error()))
+		return nil, err
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("failed to initialize AES cipher %s", err.Error()))
+		return nil, err
+	}
+
+	// Decrypt they key share, overwrite the input, and wipe the decrypted copy
+	// (input and output size are the same with AES)
+	plaintext, err := aesgcm.Open(nil, nonce, keyshare, nil)
+	if err != nil {
+		wipeKey(keyshare)
+		if plaintext != nil {
+			wipeKey(plaintext)
+		}
+		v.logger.Error(fmt.Sprintf("failed to decrypt ciphertext: %s", err.Error()))
+		return nil, err
+	}
+	wipeKey(keyshare)
```

Comment:
same comment as above, is there a reason not to always do this when we exit the function with a `defer` at the top ?

Reply:
technically not needed since this keyshare comes in encrypted, but made same change as for encrypt